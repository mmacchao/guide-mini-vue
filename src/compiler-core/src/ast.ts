export const enum NodeTypes {
    INTERPOLATION,
    SIMPLE_EXPRESSION,
    ELEMENT,
    TEXT,
    COMPUND_EXPRESSION,
    ROOT
}