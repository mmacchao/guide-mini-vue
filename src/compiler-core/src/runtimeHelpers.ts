export const TO_DISPLAY_STRING = Symbol('toDispalyString')
export const CREATE_VNODE = Symbol('createElementVnode')
export const runtimeHelpersMap = {
    [TO_DISPLAY_STRING]: 'toDisplayString',
    [CREATE_VNODE]: 'createElementVnode',
}