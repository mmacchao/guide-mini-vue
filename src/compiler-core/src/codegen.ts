import { NodeTypes } from "./ast"

export function generate(ast) {
    const node = ast.codegenNode
    const context = createCodegenContext()
    const {push} = context

    const VueBinging = 'Vue'
    const aliasHelper = s => `${s}:_${s}`
    push(`const { ${[...ast.helpers].map(aliasHelper).join(',')} } = ${VueBinging}`)
    push('\n')

    const functionName = 'render'
    const args = ['_ctx', '_cache']
    const signature = args.join(',')
    push('return function ')
    push(functionName)
    push(`(${signature})`)
    push('{ return ')
    genCode(node, context)
    push('}')

    return {
        code: context.code,
    }
}

function createCodegenContext() {
    const context = {
        code: '',
        push(str) {
            context.code += str
        }
    }
    return context
}
function genCode(node: any, context: any) {
   switch (node.type) {
    case NodeTypes.TEXT:
        genText(node, context)
        break;
    case NodeTypes.INTERPOLATION:
        genInterpolation(node, context)
        break
    case NodeTypes.SIMPLE_EXPRESSION:
        genExpression(node, context)
        break
    default:
        break;
   }
}

function genText(node, context) {
    const {push} = context
    push(`'${node.content}'`) 
}

function genInterpolation(node: any, context: any) {
    genCode(node.content, context)
}

function genExpression(node: any, context: any) {
    const {push} = context
    push(node.content) 
}

