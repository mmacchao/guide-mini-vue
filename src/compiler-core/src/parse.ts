import { NodeTypes } from "./ast"

export function baseParse(content) {
    const context = createParseContext(content)
    return createRoot(parseChildren(context))
}

function parseChildren(context) {
    const nodes: any = []
    let node
    if (context.source.startsWith('{{')) {
        node = parseInterpolation(context)
        nodes.push(node)
    } else if (context.source.startsWith('<')) {
        if (/[a-z]/i.test(context.source[1])) {
            node = parseElement(context)
        }

    }

    if(!node) {
        node = parseText(context)
    }

    nodes.push(node)

    return nodes
}
const enum TagTypes {
    START,
    END
}
function parseElement(context) {
   const node = parseTag(context, TagTypes.START)
   parseTag(context, TagTypes.END)
   return node
}
function parseTag(context, type) {
    const s = context.source
    const match: any = /^<\/?([a-z]*)/i.exec(s)
    const tag = match[1]
    // context.source = context.source.slice(match[0].length+1)
    advancedBy(context, match[0].length+1)
    if(type === TagTypes.END) return
    return {
        type: NodeTypes.ELEMENT,
        tag
    }
}

function parseInterpolation(context) {
    const openDelimiter = '{{'
    const closeDelimiter = '}}'
    const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)
    const rawContentLength = closeIndex - openDelimiter.length
    advancedBy(context, openDelimiter.length)
    const content = parseTextData(context, rawContentLength)

    return {
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: content.trim(),
        }
    }
}

function createRoot(children) {
    return {
        children: children,
    }
}

function createParseContext(content) {
    return {
        source: content
    }
}

function parseText(context: any) {
    const content = parseTextData(context, context.source.length)

    return {
        type: NodeTypes.TEXT,
        content: 'some text',
    }
}

function parseTextData(context, length) {
    const content = context.source.slice(0, length)
    advancedBy(context, content.length)
    return content
}

function advancedBy(context: any, length: any) {
    context.source = context.source.slice(length)
}

