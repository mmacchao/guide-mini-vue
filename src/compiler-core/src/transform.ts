import { NodeTypes } from "./ast"

export function transform(root, options = {}) {
    const context = createTransformContext(root, options)
    traverseNode(root, context)
    root.codegenNode = root.children[0]
    root.helpers = context.helpers.keys()
    return root
}

function createTransformContext(root: any, options: any) {
    const context: any = {
        root,
        nodeTransforms: options.nodeTransforms || [],
        helpers: new Map(),
        addHelper(helper) {
            context.helpers.set(helper, 1)
        }
    }
    return context
}
function traverseNode(node: any, context) {
    switch(node.type) {
        case NodeTypes.INTERPOLATION:
            context.addHelper('displayString')
            break
        default:
            break
    }

    context.nodeTransforms.forEach(plugin => {
        plugin(node)
    })
    if(node.children) {
        node.children.forEach(childNode => {
            traverseNode(childNode, context)
        })
    }
}

