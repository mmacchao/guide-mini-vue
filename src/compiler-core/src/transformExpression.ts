import { NodeTypes } from "./ast";

export function transformExpression(node) {
    if(node.type === NodeTypes.SIMPLE_EXPRESSION) {
        node.content = `_ctx.${node.content}`
    }
}