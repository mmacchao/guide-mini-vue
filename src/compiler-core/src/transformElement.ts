import { NodeTypes } from "./ast";
import { CREATE_VNODE, runtimeHelpersMap } from "./runtimeHelpers";

export function transfromElement(node, context) {
    if(node.type === NodeTypes.ELEMENT) {
        context.addHelper(CREATE_VNODE)
    }
}