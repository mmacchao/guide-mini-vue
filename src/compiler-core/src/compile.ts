import { generate } from "./codegen"
import { baseParse } from "./parse"
import { transform } from "./transform"
import { transfromElement } from "./transforms/transformElement"
import { transformExpression } from "./transforms/transformExpression"
import { transfromText } from "./transforms/transformText"

export function baseCompile(template) {
    const ast = baseParse(template)

    return generate(transform(ast, {
        nodeTransforms: [transformExpression, transfromElement, transfromText]
    }))
}