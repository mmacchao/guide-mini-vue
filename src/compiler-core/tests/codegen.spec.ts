import { NodeTypes } from "../src/ast"
import { generate } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"
import { transformExpression } from "../src/transformExpression"

describe('codegen', () => {
    it('string', () => {
        const content = 'hi'
        const ast = baseParse(content)

        const {code} = generate(transform(ast))
        expect(code).toMatchSnapshot()
    })

    it.only('interpolation', () => {
        const content = '{{message}}'
        const ast = baseParse(content)

        const {code} = generate(transform(ast, {
            nodeTransforms: [transformExpression]
        }))
        expect(code).toMatchSnapshot()
    })
})