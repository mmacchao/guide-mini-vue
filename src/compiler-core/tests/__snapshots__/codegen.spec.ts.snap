// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen interpolation 1`] = `
"const { displayString:_displayString } = Vue
return function render(_ctx,_cache){ return message}"
`;
